File: backend/app/services/weather_service.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import TYPE_CHECKING

import httpx
import orjson

if TYPE_CHECKING:
    from redis.asyncio import Redis as AsyncRedis

from app.core import logger
from app.core.config import get_settings
from app.core.exceptions import WeatherDomainError
from app.utils.qweather_jwt import generate_qweather_jwt

_QWEATHER_BASE_URL = get_settings().QWEATHER_BASE_URL.rstrip("/")


class _QWeatherClient:
    """Private QWeather HTTP client with auth + caching."""

    def __init__(self, redis: AsyncRedis) -> None:
        self._redis = redis

    async def get(
        self,
        path: str,
        *,
        params: dict,
        cache_key: str,
        ttl: int,
    ) -> dict:
        """Authenticated, cached GET. Raises WeatherDomainError on failure."""
        cached = await self._redis.get(cache_key)
        if cached:
            return orjson.loads(cached)

        jwt = await self._get_jwt()
        url = _QWEATHER_BASE_URL + path
        headers = {"Authorization": f"Bearer {jwt}"}

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    url, headers=headers, params=params
                )
                response.raise_for_status()
                data = response.json()
        except httpx.HTTPStatusError as exc:
            response = exc.response
            raise WeatherDomainError(
                f"QWeather API error: {response.status_code} - {response.text}",
                response.status_code,
            ) from exc
        except httpx.HTTPError as exc:
            raise WeatherDomainError(
                f"Failed to fetch QWeather data: {exc!s}", 503
            ) from exc

        await self._redis.set(cache_key, orjson.dumps(data), ex=ttl)
        return data

    def resolve_location(
        self,
        location: str | None,
        location_id: str | None,
    ) -> tuple[str, dict]:
        """Validate location input, return (value, params_dict).

        Raises WeatherDomainError if both/neither provided.
        """
        if location_id:
            return location_id, {"location": location_id}
        if location:
            return location, {"location": location}
        raise WeatherDomainError("必须提供位置信息", 400)

    async def _get_jwt(self) -> str:
        """Cached JWT generation."""
        cache_key = "qweather:jwt"
        cached_jwt = await self._redis.get(cache_key)
        if cached_jwt:
            if isinstance(cached_jwt, bytes):
                return cached_jwt.decode()
            return str(cached_jwt)

        encoded_jwt: str = generate_qweather_jwt()
        await self._redis.set(cache_key, encoded_jwt, ex=24 * 3600)
        return encoded_jwt


class WeatherService:
    async def get_qweather_tide(
        self, redis: AsyncRedis, harbor: str, date: str
    ) -> tuple[dict, bool]:
        cache_key: str = f"qweather:tide:{harbor}:{date}"
        cached = await redis.get(cache_key)
        if cached:
            return orjson.loads(cached), True

        client = _QWeatherClient(redis)
        data = await client.get(
            "/v7/ocean/tide",
            params={"location": harbor, "date": date},
            cache_key=cache_key,
            ttl=12 * 3600,
        )
        return data, False

    async def get_qweather_location(
        self, location: str, type_: str, redis: AsyncRedis
    ) -> dict:
        client = _QWeatherClient(redis)
        cache_key = f"qweather:location:{location}:{type_}"
        return await client.get(
            "/geo/v2/poi/lookup",
            params={"location": location, "type": type_},
            cache_key=cache_key,
            ttl=24 * 3600,
        )

    async def get_weather_forecast(
        self,
        days: int,
        redis: AsyncRedis,
        location: str | None = None,
        location_id: str | None = None,
    ) -> dict:
        client = _QWeatherClient(redis)
        loc_value, params = client.resolve_location(location, location_id)
        return await client.get(
            f"/v7/weather/{days}d",
            params=params,
            cache_key=f"qweather:forecast:{loc_value}:{days}d",
            ttl=3600,
        )

    async def get_current_weather(
        self,
        redis: AsyncRedis,
        location: str | None = None,
        location_id: str | None = None,
    ) -> dict:
        client = _QWeatherClient(redis)
        loc_value, params = client.resolve_location(location, location_id)
        return await client.get(
            "/v7/weather/now",
            params=params,
            cache_key=f"qweather:current:{loc_value}",
            ttl=600,
        )

    async def get_poi_lookup(self, location: str, redis: AsyncRedis) -> dict:
        client = _QWeatherClient(redis)
        return await client.get(
            "/geo/v2/poi/lookup",
            params={"location": location, "type": "scenic"},
            cache_key=f"qweather:poi:{location}:scenic",
            ttl=24 * 3600,
        )

    async def get_hourly_weather(
        self,
        hours: int,
        redis: AsyncRedis,
        location_id: str | None = None,
        location: str | None = None,
    ) -> dict:
        client = _QWeatherClient(redis)
        loc_value, params = client.resolve_location(location, location_id)
        return await client.get(
            f"/v7/weather/{hours}h",
            params=params,
            cache_key=f"qweather:hourly:{loc_value}",
            ttl=1800,
        )

    async def get_indices(
        self,
        redis: AsyncRedis,
        location: str | None = None,
        location_id: str | None = None,
    ) -> dict:
        client = _QWeatherClient(redis)
        loc_value, params = client.resolve_location(location, location_id)
        params["type"] = "4"
        return await client.get(
            "/v7/indices/1d",
            params=params,
            cache_key=f"qweather:indices:{loc_value}",
            ttl=12 * 3600,
        )

    async def get_nearby_tsta(self, location: str, redis: AsyncRedis) -> dict:
        """获取指定经纬度附近的潮汐站点（TSTA）信息。

        Returns
            dict: 包含站点名称和ID的字典，例如 {"name": "站点名称", "id": "站点ID"}。
        """
        client = _QWeatherClient(redis)
        data = await client.get(
            "/geo/v2/poi/lookup",
            params={"location": location, "type": "TSTA"},
            cache_key=f"qweather:tsta:{location}",
            ttl=24 * 3600,
        )
        poi = data.get("poi", [])[0] if data.get("poi") else {}
        return {"id": poi.get("id")}

    async def get_full_weather_data(
        self, location: str, redis: AsyncRedis
    ) -> dict:
        poi_data = await self.get_poi_lookup(location, redis)
        poi_name = (
            poi_data.get("poi", [])[0].get("name")
            if poi_data.get("poi")
            else None
        )
        poi_id = (
            poi_data.get("poi", [])[0].get("id")
            if poi_data.get("poi")
            else None
        )
        logger.info(
            f"POI lookup for location {location} returned: {poi_name} (ID: {poi_id})"
        )

        try:
            tsta_info = await self.get_nearby_tsta(location, redis)
            if tsta_info.get("id"):
                tsta_id = tsta_info["id"]
                logger.info(
                    f"Found nearby TSTA station for ID: {tsta_info['id']})"
                )
            else:
                logger.warning(
                    f"No nearby TSTA station found for location {location}"
                )
        except WeatherDomainError as exc:
            logger.error(
                f"Error fetching nearby TSTA station for location {location}: {exc!s}"
            )
            tsta_id = None

        if not poi_name and not poi_id:
            raise WeatherDomainError("无法获取位置的POI信息", 404)

        try:
            current_weather = await self.get_current_weather(
                location=location, redis=redis
            )

            hourly_weather = await self.get_hourly_weather(
                hours=24, redis=redis, location=location
            )
            daily_weather = await self.get_weather_forecast(
                location=location,
                days=3,
                redis=redis,
            )

            date_str = datetime.now(UTC).strftime("%Y%m%d")
            tide_data, _ = await self.get_qweather_tide(
                redis, harbor=tsta_id or "P2352", date=date_str
            )

            indices_data = await self.get_indices(redis, location=location)
        except WeatherDomainError as exc:
            logger.error(
                f"Error fetching weather data for location {location}: {exc!s}"
            )
            raise
        except Exception as exc:
            logger.error(
                f"Unexpected error fetching weather data for location {location}: {exc!s}",
                exc_info=True,
            )
            raise WeatherDomainError(
                "获取天气数据失败，请稍后再试", 503
            ) from exc

        logger.debug(
            f"Fetched full weather data for location: {location} (POI: {poi_name or poi_id})"
            f" - Current: {current_weather.get('now', {})}"
        )

        return {
            "current": current_weather,
            "hourly": hourly_weather,
            "daily": daily_weather,
            "tide": tide_data,
            "indices": indices_data,
            "locationName": poi_name or "",
            "poiId": poi_id or "",
        }


weather_service = WeatherService()

File: go-backend/internal/service/weather_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/KanoCifer/kuroome-blog/internal/config"
	"github.com/KanoCifer/kuroome-blog/internal/dto"
	"github.com/KanoCifer/kuroome-blog/internal/infra/httpclient"
	"github.com/KanoCifer/kuroome-blog/internal/infra/qweather"
	"github.com/KanoCifer/kuroome-blog/internal/util"
)

type Weatherer interface {
	GetTide(ctx context.Context, harbor, date string) (json.RawMessage, bool, error)

	GetCurrent(ctx context.Context, location, locationID *string) (json.RawMessage, error)
	GetHourly(ctx context.Context, hours int, location, locationID *string) (json.RawMessage, error)
	GetForecast(ctx context.Context, days int, location, locationID *string) (json.RawMessage, error)
	GetIndices(ctx context.Context, location, locationID *string) (json.RawMessage, error)
	GetPOI(ctx context.Context, location string) (json.RawMessage, error)
	GetNearbyTSTA(ctx context.Context, location string) (map[string]string, error)

	GetFullWeatherData(ctx context.Context, location string) (*dto.FullWeatherData, error)
}

// weatherPart 给 FanOut 结果打标签，配合完成顺序无关的 channel 读取。
type weatherPart struct {
	Kind string            // "poi" / "tsta" / "current" / "hourly" / "daily" / "tide" / "indices"
	Data json.RawMessage   // 通用 payload（poi / current / hourly / daily / tide / indices）
	Map  map[string]string // 仅 TSTA 使用
}

type WeatherService struct {
	qw *qweather.Client
}

func NewWeatherService(
	http *httpclient.Client,
	redis *redis.Client,
	cfg config.WeatherConfig,
	signer *qweather.Signer,
) *WeatherService {
	return &WeatherService{
		qw: qweather.NewClient(http, redis, cfg.QweatherBaseURL, signer),
	}
}

// GetTide 复用 qweather.Client 的缓存层，并在外层独立 peek redis 以
// 返回 (data, from_cache) 元组，与 Python 端行为一致。
func (s *WeatherService) GetTide(ctx context.Context, harbor, date string) (json.RawMessage, bool, error) {
	cacheKey := fmt.Sprintf("qweather:tide:%s:%s", harbor, date)

	if rdb := s.qw.Redis(); rdb != nil {
		cached, err := rdb.Get(ctx, cacheKey).Bytes()
		if err == nil && len(cached) > 0 {
			return cached, true, nil
		}
	}

	data, err := s.qw.Get(ctx, "/v7/ocean/tide",
		map[string]string{"location": harbor, "date": date},
		cacheKey, 12*time.Hour,
	)
	return data, false, err
}

func (s *WeatherService) GetCurrent(ctx context.Context, location, locationID *string) (json.RawMessage, error) {
	locValue, params, err := s.qw.ResolveLocation(location, locationID)
	if err != nil {
		return nil, err
	}
	return s.qw.Get(ctx, "/v7/weather/now", params,
		"qweather:current:"+locValue, 10*time.Minute)
}

func (s *WeatherService) GetHourly(ctx context.Context, hours int, location, locationID *string) (json.RawMessage, error) {
	locValue, params, err := s.qw.ResolveLocation(location, locationID)
	if err != nil {
		return nil, err
	}
	return s.qw.Get(ctx, fmt.Sprintf("/v7/weather/%dh", hours), params,
		"qweather:hourly:"+locValue, 30*time.Minute)
}

func (s *WeatherService) GetForecast(ctx context.Context, days int, location, locationID *string) (json.RawMessage, error) {
	locValue, params, err := s.qw.ResolveLocation(location, locationID)
	if err != nil {
		return nil, err
	}
	return s.qw.Get(ctx, fmt.Sprintf("/v7/weather/%dd", days), params,
		"qweather:forecast:"+locValue+":"+fmt.Sprintf("%dd", days), 1*time.Hour)
}

func (s *WeatherService) GetIndices(ctx context.Context, location, locationID *string) (json.RawMessage, error) {
	locValue, params, err := s.qw.ResolveLocation(location, locationID)
	if err != nil {
		return nil, err
	}
	params["type"] = "4"
	return s.qw.Get(ctx, "/v7/indices/1d", params,
		"qweather:indices:"+locValue, 12*time.Hour)
}

func (s *WeatherService) GetPOI(ctx context.Context, location string) (json.RawMessage, error) {
	return s.qw.Get(ctx, "/geo/v2/poi/lookup",
		map[string]string{"location": location, "type": "scenic"},
		fmt.Sprintf("qweather:poi:%s:scenic", location), 24*time.Hour)
}

func (s *WeatherService) GetNearbyTSTA(ctx context.Context, location string) (map[string]string, error) {
	data, err := s.qw.Get(ctx, "/geo/v2/poi/lookup",
		map[string]string{"location": location, "type": "TSTA"},
		fmt.Sprintf("qweather:tsta:%s", location), 24*time.Hour)
	if err != nil {
		return nil, err
	}
	var parsed struct {
		POI []struct {
			ID string `json:"id"`
		} `json:"poi"`
	}
	if err := json.Unmarshal(data, &parsed); err != nil {
		return nil, fmt.Errorf("weather: parse tsta: %w", err)
	}
	if len(parsed.POI) == 0 {
		return map[string]string{}, nil
	}
	return map[string]string{"id": parsed.POI[0].ID}, nil
}

func (s *WeatherService) GetFullWeatherData(ctx context.Context, location string) (*dto.FullWeatherData, error) {
	// 阶段 1：POI + TSTA 并发；POI 失败 fatal，TSTA 失败容错。
	ch1 := util.FanOut(ctx,
		func(ctx context.Context) (weatherPart, error) {
			d, err := s.GetPOI(ctx, location)
			return weatherPart{Kind: "poi", Data: d}, err
		},
		func(ctx context.Context) (weatherPart, error) {
			d, err := s.GetNearbyTSTA(ctx, location)
			return weatherPart{Kind: "tsta", Map: d}, err
		},
	)

	var poiData json.RawMessage
	var tstaID string
	for range 2 {
		var r util.Result[weatherPart]
		select {
		case r = <-ch1:
		case <-ctx.Done():
			return nil, ctx.Err()
		}
		switch r.Value.Kind {
		case "poi":
			if r.Error != nil {
				return nil, r.Error
			}
			poiData = r.Value.Data
		case "tsta":
			if r.Error != nil {
				slog.ErrorContext(ctx, "fetch nearby TSTA failed",
					"location", location, "error", r.Error.Error())
				continue
			}
			if id, ok := r.Value.Map["id"]; ok && id != "" {
				tstaID = id
				slog.InfoContext(ctx, "found nearby TSTA", "id", tstaID)
			} else {
				slog.WarnContext(ctx, "no nearby TSTA", "location", location)
			}
		}
	}

	var poiParsed struct {
		POI []struct {
			Name string `json:"name"`
			ID   string `json:"id"`
		} `json:"poi"`
	}
	if err := json.Unmarshal(poiData, &poiParsed); err != nil {
		return nil, fmt.Errorf("weather: parse poi: %w", err)
	}
	var poiName, poiID string
	if len(poiParsed.POI) > 0 {
		poiName = poiParsed.POI[0].Name
		poiID = poiParsed.POI[0].ID
	}
	slog.InfoContext(ctx, "POI lookup returned",
		"location", location, "name", poiName, "id", poiID)

	if poiName == "" && poiID == "" {
		return nil, fmt.Errorf("%w: no POI for %s", qweather.ErrUpstream, location)
	}

	dateStr := time.Now().UTC().Format("20060102")
	harbor := tstaID
	if harbor == "" {
		harbor = "P2352"
	}

	// 阶段 2：5 个 weather endpoint 并发；channel 按完成顺序到达，靠 Kind 标签归位。
	ch2 := util.FanOut(ctx,
		func(ctx context.Context) (weatherPart, error) {
			d, err := s.GetCurrent(ctx, &location, nil)
			return weatherPart{Kind: "current", Data: d}, err
		},
		func(ctx context.Context) (weatherPart, error) {
			d, err := s.GetHourly(ctx, 24, &location, nil)
			return weatherPart{Kind: "hourly", Data: d}, err
		},
		func(ctx context.Context) (weatherPart, error) {
			d, err := s.GetForecast(ctx, 3, &location, nil)
			return weatherPart{Kind: "daily", Data: d}, err
		},
		func(ctx context.Context) (weatherPart, error) {
			// GetTide 返回 (data, from_cache, error)，丢掉 cache 标记。
			d, _, err := s.GetTide(ctx, harbor, dateStr)
			return weatherPart{Kind: "tide", Data: d}, err
		},
		func(ctx context.Context) (weatherPart, error) {
			d, err := s.GetIndices(ctx, &location, nil)
			return weatherPart{Kind: "indices", Data: d}, err
		},
	)

	var current, hourly, daily, tide, indices json.RawMessage
	for range 5 {
		var r util.Result[weatherPart]
		select {
		case r = <-ch2:
		case <-ctx.Done():
			return nil, ctx.Err()
		}
		if r.Error != nil {
			return nil, r.Error
		}
		switch r.Value.Kind {
		case "current":
			current = r.Value.Data
		case "hourly":
			hourly = r.Value.Data
		case "daily":
			daily = r.Value.Data
		case "tide":
			tide = r.Value.Data
		case "indices":
			indices = r.Value.Data
		}
	}

	slog.DebugContext(ctx, "fetched full weather data",
		"location", location, "poi", poiName)

	return &dto.FullWeatherData{
		Current:      current,
		Hourly:       hourly,
		Daily:        daily,
		Tide:         tide,
		Indices:      indices,
		LocationName: poiName,
		POIID:        poiID,
	}, nil
}
